516 node ipmi control UI

steven-chiu-bigstack opened the pull request:
#### What type of PR is this?

Feat

#### What this PR does / why we need it

Implement Node IMPI control UI.

#### Which issue(s) this PR fixes

Fixes #516

#### Special notes for your reviewer

- Connecting (enabling IPMI control)

   https://github.com/user-attachments/assets/819479db-b6a9-4914-b76c-66149d6829cd

- Disconnecting

   https://github.com/user-attachments/assets/1177bbd2-5c99-47d3-8b7a-467949fae5f7

- Validation failed

https://github.com/user-attachments/assets/f8eac003-10b9-452d-a23d-7368e1ea2f5b

- New action menu items once IPMI is turned on

<img width="194" alt="{57FF9165-EE9F-4149-8AD0-09A095D294F7}" src="https://github.com/user-attachments/assets/4ed13e8a-d09e-4990-9ca0-d51a0c879c81" />

steven-chiu-bigstack committed:
refactor(web-app): rename node detail page route enum

Signed-off-by: steven-chiu-bigstack <steven.chiu@bigstack.co>

steven-chiu-bigstack committed:
feat(web-app): node impi control ui

Signed-off-by: steven-chiu-bigstack <steven.chiu@bigstack.co>

steven-chiu-bigstack committed:
feat(web-app): integrate node ipmi api

Signed-off-by: steven-chiu-bigstack <steven.chiu@bigstack.co>

steven-chiu-bigstack committed:
feat(ui-library): cos-status colors for powering on and off

Signed-off-by: steven-chiu-bigstack <steven.chiu@bigstack.co>

raven-pan commented on packages/cube-frontend-web-app/src/pages/node/[name]/_components/NodeDetailsHeader.tsx:
```diff
@@ -62,24 +103,34 @@ export const NodeDetailsHeader = (props: NodeDetailsHeaderProps) => {
+        {node?.hostname ?? ''}
+      </CosBackButton>
+      {node?.ipmi.isSupported && (
+        <Link to={CosRoutesEnum.NODE_IPMI_CONTROL_PAGE(node?.hostname)}>
```
nitpick: We already checked that `node` is not null in the previous line, so we can safely remove the `?` optional operator.

raven-pan commented on packages/cube-frontend-web-app/src/pages/node/[name]/_components/NodeSummary.tsx:
```diff
@@ -117,13 +121,21 @@ export const NodeSummary = (props: NodeSummaryProps) => {
   }
 
   const getLicenseExpiration = (): string => {
```
Maybe we could reuse `toLicenseExpirationDate` fn in `utils/date.ts`?

steven-chiu-bigstack replied:
I did think about reusing `toLicenseExpirationDate`, but `toLicenseExpirationDate` formats dates to `YYYY/MM/DD`, while [the one](https://www.figma.com/design/kRJ5MUAkMF0AYgXCeChtZn/COS-management-UI-Redesign-v.3?node-id=3281-15059&m=dev) in this page has to be `YYYY/MM/DD HH:mm`.

<img width="543" alt="{5E14DE0B-1DF2-40D8-BECA-9C6B4A7C0E02}" src="https://github.com/user-attachments/assets/59a72814-6767-4b99-acc8-da8d2c193796" />

Do you think it's a good idea to add an argument to `toLicenseExpirationDate` that determines whether to include `HH:mm` in the result? (sounds good to me!)

raven-pan replied:
Yes, Sure!  the `Unlicense` and empty string display logic seems easy to lead new developers to make mistakes.

raven-pan commented on packages/cube-frontend-web-app/src/pages/node/[name]/_components/useIPMIOperations.ts:
```diff
@@ -0,0 +1,73 @@
+    try {
+      await nodesApi.operateNodeIpmi({
+        dataCenter: dataCenter!.name,
+        nodeName: node?.hostname ?? '',
```
Discussion:
Maybe we can throw an error or early return if `nodeName` is `null` or an empty string before the API call.

steven-chiu-bigstack replied:
Absolutely. Now that you mention it, the non-null assertion operator `!` also looks reasonable to me, just like how we did it with `dataCenter!.name`. Do you think it make sense to use `node!.hostname` here? (it is to me)

raven-pan replied:
Sure!

steven-chiu-bigstack committed:
fix(web-app): node details action menu ipmi operation display conditions

Signed-off-by: steven-chiu-bigstack <steven.chiu@bigstack.co>

raven-pan commented on packages/cube-frontend-web-app/src/pages/node/[name]/ipmi-control/IPMICancelButton.tsx:
```diff
@@ -0,0 +1,22 @@
+  nodeName: string
+}
+
+export const IPMICancelButton = (props: IPMICancelButtonProps) => {
```
This component can be safely removed.

steven-chiu-bigstack replied:
Sorry, my bad. 🙏

raven-pan reviewed (commented):
The Status Component may need more width to display correctly

<img width="1442" alt="Screenshot 2025-06-30 at 6 59 39 PM" src="https://github.com/user-attachments/assets/ed39b785-c0cd-48c6-829c-75268c6ee72b" />

raven-pan reviewed (commented):
Discussion:
Should we disable the `Create Support Files` operation for power-off nodes?

<img width="1916" alt="Screenshot 2025-06-30 at 6 54 22 PM" src="https://github.com/user-attachments/assets/b4a26f1f-fa4f-4c24-917c-e22ef92f29b7" />

<img width="1920" alt="Screenshot 2025-06-30 at 6 54 02 PM" src="https://github.com/user-attachments/assets/c04ecc49-fab6-4bf9-8c50-d8196ee101e4" />

In the current implementation, the COS API will accept the operation, but no support file will be created.

https://github.com/user-attachments/assets/aee3dedc-6413-4368-9e8e-e18174e1b3c9

raven-pan reviewed (commented):
Some times the Node COS API responds with a  `""` status, which is not expected.

<img width="1916" alt="Screenshot 2025-06-30 at 7 03 59 PM" src="https://github.com/user-attachments/assets/25a24e65-cc8f-4145-9589-ff31635d5961" />

raven-pan reviewed (commented):
On the License Page, the Copy Serial Numbers modal shows `unlicensed` for powered-off Cube453 nodes. Maybe we can discuss and find a way to prevent user confusion.

<img width="2032" alt="Screenshot 2025-06-30 at 6 55 19 PM" src="https://github.com/user-attachments/assets/4984a183-f815-4ba7-bb8e-fe82b2084026" />

raven-pan reviewed (commented):
![image](https://github.com/user-attachments/assets/7757f3c7-6ce3-405e-9c8c-34b146634488)

On the IPMI Control page, the validation always passes regardless of which port the user specified. Is that expected?

We can discuss this with the COS API developers.

steven-chiu-bigstack commented:
> The Status Component may need more width to display correctly

You're right! I'll add `whitespace-nowrap` to `CosStatus` so that the text doesn't wrap:

<img width="370" alt="{80274F48-B62B-4A0A-8665-B4BD56B91C2F}" src="https://github.com/user-attachments/assets/dbcc93d6-e83f-4294-b11c-e746f00fd468" />

> Discussion:
> Should we disable the Create Support Files operation for power-off nodes?

Good catch! Create support files operation should be disabled for any node that's not in `up` status. I've also updated the Node List page (by hiding the checkboxes of non-up nodes) to align with this change.

However, the built-in disabled style from the batch action table looks a bit weird when non-text content are involved, so I adjusted  the styles to make it look slightly more natural:

![image](https://github.com/user-attachments/assets/a3ef3aaa-2736-4299-afee-59f200c9b8a6)

There'll be some edge cases where a node might be `up` when we check it, but transition to `power off` right after. I haven't handled this on the UI side because the API should always perform this check no matter what. When it happens, we can simply show a toast notification based on the API's response (and no support files will be created.)

> Some times the Node COS API responds with a "" status, which is not expected.

I've ran into this problem a few times, it's rare though. We can inform backend developers about this.

> On the License Page, the Copy Serial Numbers modal shows unlicensed for powered-off Cube453 nodes. Maybe we can discuss and find a way to prevent user confusion.

It looks like license status might be unavailable for down nodes. We can confirm this with the backend developers and see what to do next. 🙏 

> On the IPMI Control page, the validation always passes regardless of which port the user specified. Is that expected?
> We can discuss this with the COS API developers.

My guess is the validation should fail when the port is incorrect. Let's check in with the backend developers.

steven-chiu-bigstack committed:
fix(web-app): node ipmi control ui issues

Signed-off-by: steven-chiu-bigstack <steven.chiu@bigstack.co>

steven-chiu-bigstack committed:
refactor(web-app): node details page license expiration

Signed-off-by: steven-chiu-bigstack <steven.chiu@bigstack.co>

steven-chiu-bigstack committed:
style(web-app): node IPMI control page style update

Signed-off-by: steven-chiu-bigstack <steven.chiu@bigstack.co>

steven-chiu-bigstack commented:
Going to merge this PR now, really appreciate your detailed review @raven-pan! 🚀
